content/learning-paths/smartphones-and-mobile/using-neon-intrinsics-to-optimize-unity-on-android/5-the-optimizations.md

```diff
@@ -0,0 +1,275 @@
+
+## Auto-vectorization
+Producing vector-based instructions from code is called vectorization. The Burst compiler is capable of auto-vectorizing code; i.e. the programmer doesn’t have to write Neon instructions. And it can do an even better job if we follow a few simple rules (or best practice).
+
+### Best practice
+The way data and code are structured are very important factors that will influence the compiler’s ability to auto-vectorize.
+
+- Keep your code loops simple and small rather than large and complex
+
+- Where multiple data items are processed with the same operation, keep those data items together in contiguous memory (there are vector instructions for fetching, processing and saving multiple data items at once if they are in contiguous memory locations). This will also help cache efficiency.
+
+- Where the same operation is performed multiple times, try to keep those operations sequential in your code
+
+- Where possible, use fixed-length _for_ loops (they are easier for the compiler than while loops with arbitrary exit conditions)
+
+- Declare fixed-length _for_ loop counts of multiples of 4 or 8 (avoids the compiler having to produce _tail-handling_ code; i.e. code that handles the odd remaining data items)
+
+- Avoid _break_ in loops
+
+- Function calls in loops should be inlined where possible
+
+- Use simple indexing as indirect addressing can be more difficulty for the compiler to vectorise
+
+- Large data passed to a function will be passed using pointers. Those functions should be tagged with [NoAlias]. This tells the compiler that no overlaps occur in the pointers and the code is fine to be vectorized.
+
+- Don’t try to optimize every line and every function
+
+- Measure, optimize and check; iteratively make changes, measuring performance before and after to ensure each change is an improvement
+
+- Use Hint intrinsics such as _Assume_, _likely_, and _unlikely_. Read about [hints in the Burst documentation](https://docs.unity3d.com/Packages/com.unity.burst@1.8/manual/optimization-hint.html)
+
+### Modifications to collision sample to aid auto-vectorization
+The following changes were made from the plain code (unoptimized) version:
+
+1. Install Burst and use [BurstCompile] on key functions
+
+1. Collision detection routines were moved to static Burst functions
+
+1. Struct data such as AABBs stored as single array of floats
+
+1. Aggressive inlining was used by applying _[MethodImpl(MethodImplOptions.AggressiveInlining)]_ to key functions
+
+## Writing Neon intrinsics
+Before diving into the modifications made to optimize for Neon, here are some points to consider when investigating Neon for your own projects:
+
+1. The Burst compiler does a great job on it’s own but Neon intrinsics can help get a little bit more performance
```

Comment:
is this a general point regarding american english? i think in american english apostrophe isn't used so much? i ask because in some of your suggestions you do use apostrophe - are there guidelines for when apostrophes should or shouldn't be used?

Reply:
This is the guideline:

Its = what belongs to it

It's = it is

Reply:
thanks - I looked into it myself. I was completely wrong anyway - I didn't realise that even in British English you're not supposed to use an apostrophe with possessive "its"

Reply:
all fixed.
done